nodejs/src/stream_reader.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright Pravega Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use neon::prelude::*;
+use pravega_client::event::reader::{Event, SegmentSlice};
+use pravega_client::event::reader::{EventReader, EventReaderError};
+use pravega_client_shared::ScopedStream;
+use std::cell::RefCell;
+use std::sync::Arc;
+use tokio::runtime::Handle;
+use tokio::sync::Mutex;
+use tracing::info;
+
+impl Finalize for EventData {}
+
+///
+/// This represents an event that was read from a Pravega Segment and the offset at which the event
+/// was read from.
+///
+#[derive(new)]
+pub(crate) struct EventData {
+    offset_in_segment: i64,
+    value: Vec<u8>,
+}
+
+impl EventData {
+    ///
+    /// Return the event data as `ArrayBuffer`.
+    ///
+    pub fn js_data(mut cx: FunctionContext) -> JsResult<JsArrayBuffer> {
+        let event_data = cx.this().downcast_or_throw::<JsBox<EventData>, _>(&mut cx)?;
+        let data = event_data.value.as_slice().clone();
+
+        Ok(JsArrayBuffer::external(&mut cx, data.to_owned()))
+    }
+
+    ///
+    /// Return the event offset in the segment.
+    ///
+    pub fn js_offset(mut cx: FunctionContext) -> JsResult<JsNumber> {
```

Comment:
Do we want to expose this?

Reply:
Not sure. Will exposing this cause confusion to users? The Python `__repr__` implementation of `EventData` does expose the offset, so I'm assuming it's something that is ok to be exposed.

Reply:
It's fine to pass it to the application and for them to pass it back to us via APIs. But we want to be clear that one cann't do pointer math with it. IE: It is not ok to generate your own integer based on this value and then use that to refer to something, because we don't guarantee that we won't have any header or compression or something in the future. So it's fine so long as the value is not interpreted. 

In Java we do this by having an Object which wraps it which is opaque to users, so while it is just a long they can't directly get that long for visibility reasons. This is the normal pattern in Java to lock things down. In Python that is not the normal pattern and underscores are used in the name to indicate caution. I don't know what the norm is in JS. Is there a common pattern around this?

Reply:
There is a keyword called `readonly` that works exactly as Python's underscore pattern. But if you've written Python or JavaScript or any similar dynamic code, you would know users could access and manipulate anything they want. So instead of posing some restrictions on offset, it may be better to state it in doc or comment and let users be clear about what they are doing.

Besides, both py and js bindings don't allow users to specify a streamcut/offset on reading, so we may also improve this later.